(WIP) Revision of schema to JSON Schema draft 7

Mitriol opened the pull request:
The schema output by "inspec schema <name>" has been revised to be in the JSON Schema draft 7 format, providing better support for required attributes and self-documenting schema structures. As of right now this provides no functional change to what the schema _is_ (the reporters are completely untouched), simply how it is generated and output.

However, this work should hopefully be helpful to developers hoping to understand/automatically parse the Inspec output schema, and also to be a good foundation for any future revisions/versioning of the schema.

## Related Issue
See #4282 , #3706 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the **CONTRIBUTING** document.

## Future work:
- [X] Update existing schema structure to draft 7
- [X] Update tests to validate new schema version (was already done)
- [X] Make tests validate the schema itself
- [ ] Add descriptions to all fields (where appropriate)
- [X] Verify that all required data is marked required (and, conversely, that no non-required data is marked required).
- [ ] Document the current schema generation process

And, optionally, the following:
- [ ] Add versioning component/parameter to the schema generation cli, to support any future breaking changes
- [ ] Unify schema types to avoid inconsistent data (e.g. the description labels in #3682).
- [ ] Decide how to version schema generation to accommodate future changes. Should json schema <version> simply fetch a cached output of schema <version> if it is not the newest version?

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Initial commit of changes

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Moved Control Description/Result to exec_json

Mitriol committed:
Made SchemaTypes not default-object

Mitriol committed:
Enumerated control result statuses

miah commented on lib/inspec/schema/profile_json.rb:
```diff
@@ -0,0 +1,58 @@
+require_relative "./primitives"
```
`require "inspec/schema/primitives"`

miah commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+require_relative "./primitives"
+require_relative "./exec_json"
+require_relative "./exec_json_min"
+require_relative "./profile_json"
```
Also here, I don't think we need require_relative as require will work fine for these.

Mitriol replied:
Changed all requires to `require "lib/inspec/schema/__file__"`

miah commented on lib/inspec/schema/profile_json.rb:
```diff
@@ -0,0 +1,58 @@
+require_relative "./primitives"
+
+# We disable this for the sake of legibility
+# rubocop:disable Layout/AlignHash
```
We should avoid adding new rubocop:disables to the code.

Mitriol replied:
Done. Having them aligned was nice for working, but unnecessary for dev.

miah commented on lib/inspec/schema/exec_json.rb:
```diff
@@ -0,0 +1,73 @@
+    # Based loosely on https://www.inspec.io/docs/reference/profiles/ as of July 3, 2019
+    # However, concessions were made to the reality of current reporters, specifically
+    # with how description is omitted and version/inspec_version aren't as advertised online
+    PROFILE = SchemaType.new("Exec JSON Profile", {
```
even if version/inspec_version aren't as advertised we should still include them.

Mitriol replied:
Done

miah reviewed (changes requested):
Thanks @Mitriol!

Mitriol committed:
Small naming, formatting tweaks

Mitriol committed:
Updated "require"s to be non-relative (re @miah)

Mitriol committed:
Removed rubocop ignores, ran chefstyle.

No other changes beyond style were made in this commit.

zenspider commented on lib/inspec/schema/exec_json.rb:
```diff
@@ -1,4 +1,4 @@
-require_relative "./primitives"
+require "lib/inspec/schema/primitives"
```
no, not "lib". lib is already in the load path (via rake/gemfile/etc)... this would only work if "." were in the load path as well, which it isn't by default.

Mitriol replied:
Resolved as of https://github.com/inspec/inspec/pull/4330/commits/6dd9b4d21f8db0c4d8180795c6059f6fedb08d31.

Seemed to be working either way but yes, this is nicer.

zenspider commented on lib/inspec/schema/exec_json.rb:
```diff
@@ -1,66 +1,63 @@
+        "message" => STRING,
+        "skip_message" => STRING,
+        "exception" => STRING,
+        "backtrace" => OBJECT,
```
Do we really have a whitespace linter rule turned on? @miah?

I will fight chefstyle on this one. Visual pattern matching is a good thing to the point that we're literally evolved for it. We shouldn't make it harder.

zenspider commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+require "json"
```
this file should be named `lib/inspec/schema/output_schema.rb`, no?

Mitriol replied:
Resolved as of https://github.com/inspec/inspec/pull/4330/commits/a25eabb6d2b7289edc3fccd6771b17767e2fc306

zenspider commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+require "lib/inspec/schema/profile_json"
+
+module Inspec
+  module OutputSchema
```
and should be namespaced to match.

Mitriol replied:
If I'm being fully honest I'd never written ruby until ~2 weeks ago, so still not entirely clear on some best-practice stuff. 

To be clear, are you saying it should be
`
namespace OutputSchema
    module OutputSchema
`

Apologies if I'm being obtuse here.

miah replied:
```
module Inspec
  module Schema
    module OutputSchema
    end
  end
end
```

Mitriol replied:
Resolved as of https://github.com/inspec/inspec/pull/4330/commits/e7cb80d927b114bcbb0a3627dba84805f2be3300

zenspider commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+      # Nest under "definitions" label
+      {
+        "definitions" => definitions,
+      }
```
All of the comments are redundant distractions. Each "paragraph" of code is separated but actually all related. And all boil down to something that's both succinct and easily comprehensible:

```ruby
{
  "definitions" => schema_type.all_depends.map { |t| [t.name, t.body] }.to_h
}
```

Mitriol replied:
Used your change in https://github.com/inspec/inspec/pull/4330/commits/7fe53fc1fc53ba04e5579e5542011d7d5f61bba1

zenspider commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+      "profile-json" => OutputSchema.finalize(Schema::ProfileJson::PROFILE),
+      "exec-json" => OutputSchema.finalize(Schema::ExecJson::OUTPUT),
+      "exec-jsonmin" => OutputSchema.finalize(Schema::ExecJsonMin::OUTPUT),
+      "platforms" => PLATFORMS,
```
why isn't the proc here?

zenspider replied:
Also, do we benefit from the lazy aspect at all? I can't tell when/where/how "pratforms" gets hit.

Mitriol replied:
Forgive me but what is the proc?

RE: Lazy aspect: Platforms is unchanged from the old code - I figured there must have been a reason for it at the time, but I can remove it now if necessary.

zenspider commented on lib/inspec/schema/schema.rb:
```diff
@@ -0,0 +1,64 @@
+
+    # using a proc here so we can lazy load it when we need
+    # TODO: Modernize
+    PLATFORMS = lambda do
```
Why the TODO? Modernize in what sense? Why isn't it just done the way you want?

aaronlippold replied:
I think a few of these things are historical, as we were trying to be as non-invasive as possible will still achieving our initial goal. We'll go through and remove any such comments of actions to be taken and replace them with any notes of things that still need to be done.

Mitriol replied:
^ I don't know what modernize means here either haha.

zenspider commented on lib/inspec/schema/profile_json.rb:
```diff
@@ -0,0 +1,58 @@
+# These type occur only when running "inspec json <file>".
+
+module Schema
+  module ProfileJson
```
namespacing should match pathing.

Mitriol replied:
Resolved as of https://github.com/inspec/inspec/pull/4330/commits/e7cb80d927b114bcbb0a3627dba84805f2be3300

zenspider commented on lib/inspec/schema/exec_json.rb:
```diff
@@ -1,66 +1,63 @@
 require "lib/inspec/schema/primitives"
 
-# We disable this for the sake of legibility
-# rubocop:disable Layout/AlignHash
```
Honestly, I could do without this commit as a whole. I think it looks/works better before.

THAT MIGHT JUST BE ME THO. I haven't had this fight with the ppl owning chefstyle (yet).

miah replied:
I think we should have the chefstyle discussion, but for now I'm not a fan of adding more `# rubocop:disable`. I'd like to remove all of them from the code and just let the style system manage it while we manage the style itself.

Mitriol replied:
On the docs page for this styling rule: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Layout/AlignHash

It seems that the configuration `EnforcedHashRocketStyle: table` would result in styling as I had it, that would be manage automatically by rubocop. I'll leave this commit as is for now though.

zenspider reviewed (changes requested):
Thank you for the PR. Some changes requested.

Also, I thought based on our discussion that you wanted to go to v4, not v7?

Mitriol committed:
Update module structure to match directory paths

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Removed rubocop disable on primitives.rb

Unsure how I missed this before, but here we are.

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Simplified require paths

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Renamed schema.rb -> output_schema.rb

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Simplified definition code

(Thanks @zenspider!)

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Merge branch 'master' of https://github.com/inspec/inspec into jh/schema-improvements

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Ran chefstyle

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Moved schema validation to its own function.

This should cut down on initializer complexity.

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added some schema generation tests

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Further simplified schema validation code.

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Fixed code climate issues

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Adjusted schema validation again.

This now fails codeclimate, but is far more legible in my opinion.
Perhaps the complexity requirements are overly harsh?

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added jsons schema validation to "inspec json"

Also added a file (minimal.rb) to test that validation of
a "minimal" control (one with no data beside name/impact)
works properly.

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Chefstyle fix

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added test that verifies that all test profiles match schema

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Fixed backtrace to be an array of strings

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added tests to validate EVERY profile in the test directory.

Note that this is probably overkill, and fairly slow.

But it does make sure our validators are not over-strict!

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Removed old schema.rb

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Cleanup

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Tidied up test code

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Fixed profile paths (oops)

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Fixed reference pointers/URI vals to not have whitespace.

Prior implementation was technically not valid URI.

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Moved to a more modern/actually working schema validator gem.

The previous one did not actually validate our schema, and was
allowing many errors to slip through. This one should perform much
better (and has already! See next commit)

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added dependencies, parent profiles, small tweaks

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added some properties neglected because of additionalProperties

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Fixed missing schema dependency

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added missing branch property to dependencies. Also allowed additional properties due to uncertainty

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Made schemas less strict. Updated readme

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Merge branch 'master' of github.com:mitre/inspec into jh/schema-improvements

Just bringing things up to date

Mitriol committed:
Further improvements and relaxations on schema

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Added missing 'version' from profile json output

Signed-off-by: Jacob Henry <jchenry@mitre.org>

Mitriol committed:
Pull up to date with master

Signed-off-by: Jacob Henry <jchenry@mitre.org>
